File: Algorithms/Javascript/quickSort.js
/*
**************** Quick Sort ******************
1. Like merge sort, exploits the fact that arrays of 0 or 1 element are always sorted
2. Works by selecting one element(called the 'pivot') and finding the index where the pivot shoud end up in the sorted array
3. Once the pivot is positioned appropriately, quick sort can be applied on either side of the pivot
*/

/*
Pseudo Code
1. Call the pivot helper on the array
2. When the helper returns to you the updated pivot index, recursively call the pivot helper on the subarray to the left of that index, and the subarray to the right of that index
*/

function pivot(arr, start = 0, end = arr.length - 1) {
  const swap = (arr, idx1, idx2) => {
    [arr[idx1], arr[idx2]] = [arr[idx2], arr[idx1]];
  };

  // We are assuming the pivot is always the first element
  let pivot = arr[start];
  let swapIdx = start;

  for (let i = start + 1; i <= end; i++) {
    if (pivot > arr[i]) {
      swapIdx++;
      swap(arr, swapIdx, i);
    }
  }

  // Swap the pivot from the start the swapPoint
  swap(arr, start, swapIdx);
  return swapIdx;
}

function quickSort(arr, left = 0, right = arr.length - 1) {
  if (left < right) {
    let pivotIndex = pivot(arr, left, right); //3
    //left
    quickSort(arr, left, pivotIndex - 1);
    //right
    quickSort(arr, pivotIndex + 1, right);
  }
  return arr;
}

quickSort([100, -3, 2, 4, 6, 9, 1, 2, 5, 3, 23]);

// Time Complexity =======> O(n log n)
//  Space Complexity ======> O(log n)

File: Data Structures/Searching and Sorting/QuickSort.java
public class QuickSort {

    public void quicksort(int A[], int low, int high) {
        if(low<high) {
            int pi = partition(A,low,high);
            quicksort(A,low,pi-1);
            quicksort(A,pi+1,high);
        }
    }

    public int partition(int A[], int low, int high) {
        int pivot = A[low];
        int i = low+1;
        int j = high;
        do {
            while(i <= j && A[i] <= pivot)
                i = i + 1;
            while (i <= j && A[j] > pivot)
                j = j - 1;
            if (i <= j)
                swap(A,i,j);
        } while(i<j);
        swap(A,low,j);
        return j;
    }

    public void swap(int A[], int i, int j) {
        int temp = A[i];
        A[i] = A[j];
        A[j] = temp;
    }
    public void display(int A[], int n) {
        for(int i=0;i<n;i++)
            System.out.print(A[i] + " ");
        System.out.println();
    }

    public static void main(String args[]) {
        QuickSort s = new QuickSort();
        int A[] = {5, 3, 8, 9, 4, 6};
        System.out.print("Original Array: ");
        s.display(A, 6);
        s.quicksort(A, 0, A.length-1);
        System.out.print("Sorted Array: ");
        s.display(A, 6);
    }

}
